File: ruby/spec/rock_paper_scissors_spec.rb
require 'rspec'
require_relative '../lib/rock_paper_scissors'

describe 'Rock-Paper-Scissors' do

  let(:player1) { Player.new 'A' }
  let(:player2) { Player.new 'B' }
  let(:game) { Game.new player1, player2 }

  context 'winning and losing' do

    describe 'rock' do

      it 'should beat scissors' do
        player1.picks :rock and player2.picks :scissors
        expect(game.winner).to be(player1)
      end

      it 'should lose to paper' do
        player1.picks :rock and player2.picks :paper
        expect(game.winner).to be(player2)
      end

    end

    describe 'paper' do

      it 'should beat rock' do
        player1.picks :paper and player2.picks :rock
        expect(game.winner).to be(player1)
      end

      it 'should lose to scissors' do
        player1.picks :paper and player2.picks :scissors
        expect(game.winner).to be(player2)
      end

    end

    describe 'scissors' do

      it 'should beat paper' do
        player1.picks :scissors and player2.picks :paper
        expect(game.winner).to be(player1)
      end

      it 'should lose to rock' do
        player1.picks :scissors and player2.picks :rock
        expect(game.winner).to be(player2)
      end

    end

  end

  context 'draws' do

    describe 'any identical picks' do

      it 'should result in no winner' do

        draw_game_results = [:rock, :paper, :scissors].map do |x|
          player1.picks x and player2.picks x
          game.winner
        end

        expect(draw_game_results).to eq [nil, nil, nil]

      end

    end

  end

end

File: app/public/javascript/spec/RockPaperScissorsSpec.js
describe("Rock-Paper-Scissors", function() {
  var player1, player2, game;

  beforeEach(function() {

    player1 = new Player('Ross');
    player2 = new Player('Alex');
    game = new Game(player1, player2);

  });

  describe('winning and losing', function() {

    describe('rock', function() {

      it('should beat scissors', function() {

        player1.picks('rock');
        player2.picks('scissors');
        expect(game.winner()).toBe(player1);
        // expect(game.winnerMessage()).toBe('Rock crushes scissors');

      });

      it('should beat lizard', function() {

        player1.picks('rock');
        player2.picks('lizard');
        expect(game.winner()).toBe(player1);
        expect(game.winnerMessage()).toBe('rock crushes lizard');

      });

      it('should lose to paper', function() {

        player1.picks('rock');
        player2.picks('paper');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('paper covers rock');

      });

      it('should lose to spock', function() {

        player1.picks('rock');
        player2.picks('spock');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('spock vaporizes rock');

      });

    });

    describe('paper', function() {

      it('should beat rock', function() {

        player1.picks('paper');
        player2.picks('rock');
        expect(game.winner()).toBe(player1);
        expect(game.winnerMessage()).toBe('paper covers rock');

      });

      it('should beat spock', function() {

        player1.picks('paper');
        player2.picks('spock');
        expect(game.winner()).toBe(player1);
        expect(game.winnerMessage()).toBe('paper disproves spock');

      });

      it('should lose to scissors', function() {

        player1.picks('paper');
        player2.picks('scissors');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('scissors cut paper');


      });

      it('should lose to lizard', function() {

        player1.picks('paper');
        player2.picks('lizard');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('lizard eats paper');

      });

    });

    describe('scissors', function() {

      it('should beat paper', function() {

        player1.picks('scissors');
        player2.picks('paper');
        expect(game.winner()).toBe(player1);
        expect(game.winnerMessage()).toBe('scissors cut paper');

      });

      it('should beat lizard', function() {

        player1.picks('scissors');
        player2.picks('lizard');
        expect(game.winner()).toBe(player1);
        expect(game.winnerMessage()).toBe('scissors decapitate lizard')

      });

      it('should lose to rock', function() {

        player1.picks('scissors');
        player2.picks('rock');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('rock breaks scissors');

      });

      it('should lose to spock', function() {

        player1.picks('scissors');
        player2.picks('spock');
        expect(game.winner()).toBe(player2);
        expect(game.winnerMessage()).toBe('spock smashes scissors');

      });
    });
  });
  
  describe('winning message', function() {
    
    describe('rock when wins', function(){

      it('against scissors', function() {
        player1.picks('rock');
        player2.picks('scissors');
        expect(game.winnerMessage()).toBe('rock breaks scissors');
      })
    })
  })

  describe('draws', function() {

    describe('any identical picks', function() {

      it('should result in no winner', function() {

        var drawGameResults = ['rock', 'paper', 'scissors', 'lizard', 'spock'].map(function(x) {
          player1.picks(x);
          player2.picks(x);
          return game.winner();
        });
        expect(drawGameResults).toEqual([null, null, null, null, null]);
      });
    });
  });
});